File: NET_project/Benchmarks/FirefliesFlocking.cs
using System.Numerics;
using System.Runtime.CompilerServices;

namespace NET_project.Benchmarks;

internal struct Boid {
	public Vector3 position, velocity, acceleration;
}

public unsafe struct FirefliesFlockingNET : IJob {
	public uint boids;
	public uint lifetime;
	public float result;

	public void Run() {
		result = FirefliesFlocking(boids, lifetime);
	}

	private static uint parkMiller;
	private static float maxSpeed;
	private static float maxForce;
	private static float separationDistance;
	private static float neighbourDistance;

	private static float FirefliesFlocking(uint boids, uint lifetime) {
		parkMiller = 666;
		maxSpeed = 1.0f;
		maxForce = 0.03f;
		separationDistance = 15.0f;
		neighbourDistance = 30.0f;

		Boid[] fireflies = new Boid[boids];

		for (uint i = 0; i < fireflies.Length; i++) {
			ref Boid firefly = ref fireflies[i];

			firefly.position = new Vector3(Random(), Random(), Random());
			firefly.velocity = new Vector3(Random(), Random(), Random());
			firefly.acceleration = Vector3.Zero;
		}

		for (uint i = 0; i < lifetime; i++)
		{
			// Update
			for (uint boid = 0; boid < fireflies.Length; boid++)
			{
				ref Boid firefly = ref fireflies[boid];
				firefly.velocity += firefly.acceleration;

				float speed = firefly.velocity.Length();

				if (speed > maxSpeed)
				{
					firefly.velocity /= speed;
					firefly.velocity *= maxSpeed;
				}

				firefly.position += firefly.velocity;
				firefly.acceleration *= maxSpeed;
			}

			// Separation
			for (uint boid = 0; boid < fireflies.Length; boid++)
			{
				ref Boid firefly = ref fireflies[boid];

				Vector3 separation = Vector3.Zero;
				int count = 0;

				for (uint target = 0; target < fireflies.Length; target++) 
				{
					Vector3 position = firefly.position;

					position -= fireflies[target].position;

					float distance = position.Length();

					if (distance > 0.0f && distance < separationDistance)
					{
						position = Vector3.Normalize(position);
						position /= distance;

						separation = position;
						count++;
					}
				}

				if (count > 0)
				{
					separation /= count;
					separation = Vector3.Normalize(separation);
					separation *= maxSpeed;
					separation -= firefly.velocity;

					float force = separation.Length();

					if (force > maxForce)
					{
						separation /= force;
						separation *= maxForce;
					}

					separation *= 1.5f;
					firefly.acceleration += separation;
				}
			}

			// Cohesion
			for (uint boid = 0; boid < fireflies.Length; boid++)
			{
				ref Boid firefly = ref fireflies[boid];

				Vector3 cohesion = Vector3.Zero;
				int count = 0;

				for (uint target = 0; target < fireflies.Length; target++)
				{
					Vector3 position = firefly.position;

					position -= fireflies[target].position;

					float distance = position.Length();

					if (distance > 0.0f && distance < neighbourDistance)
					{
						cohesion = firefly.position;
						count++;
					}
				}

				if (count > 0)
				{
					cohesion /= count;
					cohesion -= firefly.position;
					cohesion = Vector3.Normalize(cohesion);
					cohesion *= maxSpeed;
					cohesion -= firefly.velocity;

					float force = cohesion.Length();

					if (force > maxForce)
					{
						cohesion /= force;
						cohesion *= maxForce;
					}

					firefly.acceleration += cohesion;
				}
			}
		}

		return parkMiller;
	}

	[MethodImpl(MethodImplOptions.AggressiveInlining)]
	private static float Random() {
		parkMiller = (uint)((ulong)parkMiller * 48271u % 0x7fffffff);

		return parkMiller / 10000000.0f;
	}
}

internal unsafe struct FirefliesFlockingGCC : IJob {
	public uint boids;
	public uint lifetime;
	public float result;
	public string libName;

	public void Run() {
		switch (libName)
		{
			case "GCC":
				result = GccNativeBindings.benchmark_fireflies_flocking(boids, lifetime);
				break;

			case "Clang":
				result = ClangNativeBindings.benchmark_fireflies_flocking(boids, lifetime);
				break;

			case "MS":
				result = MsNativeBindings.benchmark_fireflies_flocking(boids, lifetime);
				break;

			case "Rust":
				result = RustNativeBindings.benchmark_fireflies_flocking(boids, lifetime);
				break;
		}
	}
}

File: rust_lib/src/f32_benchmarks/FirefliesFlocking.rs


#[derive(Clone, Copy)]
struct Vector
{
    x: f32,
    y: f32,
    z: f32,
}

#[derive(Clone, Copy)]
struct Boid {
    position: Vector,
    velocity: Vector,
    acceleration: Vector,
}


fn benchmark_fireflies_flocking_add(left: &mut Vector, right: &mut Vector)
{
    left.x += right.x;
    left.y += right.y;
    left.z += right.z;
}

fn benchmark_fireflies_flocking_divide(vector: &mut Vector, value: f32)
{
    vector.x /= value;
    vector.y /= value;
    vector.z /= value;
}

fn benchmark_fireflies_flocking_length(vector: Vector) -> f32
{
    return f32::sqrt(vector.x * vector.x + vector.y * vector.y + vector.z * vector.z);
}

fn benchmark_fireflies_flocking_multiply(vector: &mut Vector, value: f32)
{
    vector.x *= value;
    vector.y *= value;
    vector.z *= value;
}

fn benchmark_fireflies_flocking_normalize(vector: &mut Vector)
{
    let length: f32 = f32::sqrt(vector.x * vector.x + vector.y * vector.y + vector.z * vector.z);

    vector.x /= length;
    vector.y /= length;
    vector.z /= length;
}

fn benchmark_fireflies_flocking_random(parkMiller: &mut u32) -> f32
{
    *parkMiller = (((*parkMiller as u64) * 48271) % 0x7fffffff) as u32;

    return *parkMiller as f32 / 10000000.0;
}

fn benchmark_fireflies_flocking_subtract(left: &mut Vector, right: Vector)
{
    left.x -= right.x;
    left.y -= right.y;
    left.z -= right.z;
}

pub(crate) fn benchmark_fireflies_flocking_main(boids: usize, lifetime: usize) -> f32
{
    let mut i: usize;
    let mut parkMiller: u32 = 666;
    let maxSpeed: f32 = 1.0;
    let maxForce: f32 = 0.03;
    let separationDistance: f32 = 15.0;
    let neighbourDistance: f32 = 30.0;

    let mut fireflies: Vec<Boid> = Vec::with_capacity(boids);

    i = 0;
    while (i < boids)
    {
        let firefly: Boid = Boid {
            position: Vector {
                x: benchmark_fireflies_flocking_random(&mut parkMiller),
                y: benchmark_fireflies_flocking_random(&mut parkMiller),
                z: benchmark_fireflies_flocking_random(&mut parkMiller)
            },
            velocity: Vector {
                x: benchmark_fireflies_flocking_random(&mut parkMiller),
                y: benchmark_fireflies_flocking_random(&mut parkMiller),
                z: benchmark_fireflies_flocking_random(&mut parkMiller)
            },
            acceleration: Vector {
                x: 0.0,
                y: 0.0,
                z: 0.0,
            },
        };
        fireflies.push(firefly);

        i += 1;
    }

    i = 0;
    while (i < lifetime)
    {
        // Update
        let mut boid: usize = 0;
        while (boid < boids)
        {
            let mut fireflyVelocity: Vector = fireflies[boid].velocity;
            let mut fireflyAcceleration: Vector = fireflies[boid].acceleration;
            benchmark_fireflies_flocking_add(&mut fireflyVelocity, &mut fireflyAcceleration);
            fireflies[boid].velocity = fireflyVelocity;
            fireflies[boid].acceleration = fireflyAcceleration;

            let speed: f32 = benchmark_fireflies_flocking_length(fireflies[boid].velocity);

            if (speed > maxSpeed) {
                benchmark_fireflies_flocking_divide(&mut fireflies[boid].velocity, speed);
                benchmark_fireflies_flocking_multiply(&mut fireflies[boid].velocity, maxSpeed);
            }

            fireflyVelocity = fireflies[boid].velocity;
            let mut fireflyPosition: Vector = fireflies[boid].position;
            benchmark_fireflies_flocking_add(&mut fireflyPosition, &mut fireflyVelocity);
            fireflies[boid].velocity = fireflyVelocity;
            fireflies[boid].position = fireflyPosition;

            benchmark_fireflies_flocking_multiply(&mut fireflies[boid].acceleration, maxSpeed);

            boid += 1;
        }

        // Separation
        boid = 0;
        while (boid < boids)
        {
            let mut separation: Vector = Vector { x: 0.0, y: 0.0, z: 0.0 };
            let mut count: i32 = 0;

            let mut target: usize = 0;
            while (target < boids)
            {
                let mut position: Vector = fireflies[boid].position;

                benchmark_fireflies_flocking_subtract(&mut position, fireflies[target].position);

                let distance: f32 = benchmark_fireflies_flocking_length(position);

                if (distance > 0.0 && distance < separationDistance)
                {
                    benchmark_fireflies_flocking_normalize(&mut position);
                    benchmark_fireflies_flocking_divide(&mut position, distance);

                    separation = position;
                    count += 1;
                }

                target += 1;
            }

            if (count > 0)
            {
                benchmark_fireflies_flocking_divide(&mut separation, count as f32);
                benchmark_fireflies_flocking_normalize(&mut separation);
                benchmark_fireflies_flocking_multiply(&mut separation, maxSpeed);
                benchmark_fireflies_flocking_subtract(&mut separation, fireflies[boid].velocity);

                let force: f32 = benchmark_fireflies_flocking_length(separation);

                if (force > maxForce) {
                    benchmark_fireflies_flocking_divide(&mut separation, force);
                    benchmark_fireflies_flocking_multiply(&mut separation, maxForce);
                }

                benchmark_fireflies_flocking_multiply(&mut separation, 1.5);
                benchmark_fireflies_flocking_add(&mut fireflies[boid].acceleration, &mut separation);
            }

            boid += 1;
        }

        // Cohesion
        boid = 0;
        while (boid < boids)
        {
            let mut cohesion: Vector = Vector { x: 0.0, y: 0.0, z: 0.0 };
            let mut count: i32 = 0;

            let mut target: usize = 0;
            while (target < boids)
            {
                let mut position: Vector = fireflies[boid].position;

                benchmark_fireflies_flocking_subtract(&mut position, fireflies[target].position);

                let distance: f32 = benchmark_fireflies_flocking_length(position);

                if (distance > 0.0 && distance < neighbourDistance)
                {
                    cohesion = fireflies[boid].position;
                    count += 1;
                }

                target += 1;
            }

            if (count > 0)
            {
                benchmark_fireflies_flocking_divide(&mut cohesion, count as f32);
                benchmark_fireflies_flocking_subtract(&mut cohesion, fireflies[boid].position);
                benchmark_fireflies_flocking_normalize(&mut cohesion);
                benchmark_fireflies_flocking_multiply(&mut cohesion, maxSpeed);
                benchmark_fireflies_flocking_subtract(&mut cohesion, fireflies[boid].velocity);

                let force: f32 = benchmark_fireflies_flocking_length(cohesion);

                if (force > maxForce)
                {
                    benchmark_fireflies_flocking_divide(&mut cohesion, force);
                    benchmark_fireflies_flocking_multiply(&mut cohesion, maxForce);
                }

                benchmark_fireflies_flocking_add(&mut fireflies[boid].acceleration, &mut cohesion);
            }

            boid += 1;
        }

        i += 1;
    }

    return parkMiller as f32;
}
